pkg/syncer/spec/spec_process.go

```diff
@@ -181,17 +199,29 @@ func (c *Controller) ensureDownstreamNamespaceExists(ctx context.Context, downst
 		})
 	}
 
-	// TODO(sttts): check that namespace exists in lister before using the client
-	if _, err := namespaces.Create(ctx, newNamespace, metav1.CreateOptions{}); err != nil {
-		// An already exists error is ok - it means something else beat us to creating the namespace.
-		if !k8serrors.IsAlreadyExists(err) {
-			// Any other error is not good, though.
-			// TODO bubble this up as a condition somewhere.
-			klog.Errorf("Error while creating namespace %q: %v", downstreamNamespace, err)
+	// Check if the namespace already exists, if not create it.
+	namespace, err := c.getDownstreamNamespace(newNamespace.GetName())
+	if apierrors.IsNotFound(err) {
+		// Not found, so let's create the namespace.
+		if _, err := namespaces.Create(ctx, newNamespace, metav1.CreateOptions{}); err != nil {
 			return err
 		}
-	} else {
-		klog.Infof("Created downstream namespace %s for upstream namespace %s|%s", downstreamNamespace, l.LogicalCluster, l.Namespace)
+		klog.Infof("Created downstream namespace %s for upstream namespace %s|%s", newNamespace.GetName(), desiredNSLocator.Workspace, desiredNSLocator.Namespace)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// The namespace exists, so check if it has the correct namespace locator.
+	nsLocator, exists, err := shared.LocatorFromAnnotations(namespace.GetAnnotations())
+	if err != nil {
+		return fmt.Errorf("(possible namespace collision) namespace %s already exists, but found an error when trying to decode the annotation: %w", newNamespace.GetName(), err)
+	}
+	if !exists {
+		return fmt.Errorf("(namespace collision) namespace %s has no namespace locator", namespace.GetName())
```

Comment:
this will hotloop, won't it?

Reply:
Log error and return nil? There's some way to bubble up the error somewhere visible? as the user feedback will be zero if not...

Reply:
If we created another namespace with the same nslocator, we will hit: https://github.com/jmprusi/kcp/blob/jmprusi/reworking-namespaces/pkg/syncer/spec/spec_process.go#L13